store/sqlstore/upgrade.go

```diff
@@ -1000,6 +1001,99 @@ func upgradeDatabaseToVersion533(sqlStore *SqlStore) {
+
+	// loop over channels and update the memberships for each
+	var channels []struct {
+		Id                string
+		LastPostAt        int64
+		TotalMsgCountRoot int64
+	}
+	t1 = time.Now()
+	if _, err := sqlStore.GetMaster().Select(&channels, "SELECT Id,TotalMsgCountRoot,LastPostAt FROM Channels"); err != nil {
+		mlog.Error("Error fetching Channels table", mlog.Err(err))
+	}
+	mlog.Info("Fetched Channels", mlog.String("took", time.Now().Sub(t1).String()))
+	gt1 := time.Now()
+	for _, channel := range channels {
+		t1 := time.Now()
+		// update MsgCountRoot for up to date memberships
+		if _, err := sqlStore.GetMaster().Exec("UPDATE ChannelMembers SET MsgCountRoot = :TotalMsgCountRoot WHERE ChannelId=:ChannelId AND LastViewedAt >= :LastPostAt", map[string]interface{}{"ChannelId": channel.Id, "LastPostAt": channel.LastPostAt, "TotalMsgCountRoot": channel.TotalMsgCountRoot}); err != nil {
+			mlog.Error("Error updating MsgCountRoot for up-to-date memberships", mlog.Err(err))
+			break
+		}
+		// collect all memberships that need update
+		var memberships []struct {
+			UserId       string
+			LastViewedAt int64
+		}
+		if _, err := sqlStore.GetMaster().Select(&memberships, "SELECT UserId, LastViewedAt FROM ChannelMembers WHERE  ChannelId=:ChannelId ORDER BY LastViewedAt DESC", map[string]interface{}{"ChannelId": channel.Id}); err != nil {
+			mlog.Error("Error getting channelMemberships", mlog.Err(err))
+			break
+		}
+		// calculate oldest 'viewedAt' for memberships and use that value to filter out some of the posts
+		minLastViewedAt := int64(math.MaxInt64)
+		for _, m := range memberships {
+			if m.LastViewedAt < minLastViewedAt {
+				minLastViewedAt = m.LastViewedAt
+			}
+		}
+		var postTimestamps []int64
+		if _, err := sqlStore.GetMaster().Select(&postTimestamps, "SELECT UpdateAt FROM Posts WHERE ChannelId=:ChannelId AND DeleteAt=0 AND RootId='' AND CreateAt > :MinLastViewedAt ORDER BY CreateAt DESC", map[string]interface{}{"ChannelId": channel.Id, "MinLastViewedAt": minLastViewedAt}); err != nil {
+			mlog.Error("Error gettings posts", mlog.Err(err))
+			break
+		}
+		membershipIndex := 0
+		transaction, err := sqlStore.GetMaster().Begin()
+		if err != nil {
+			mlog.Error("Error starting transaction", mlog.Err(err))
+			break
+		}
+		defer finalizeTransaction(transaction)
+
+		for counter, postTimestamp := range postTimestamps {
+			for {
+				if counter == 0 || len(memberships) <= membershipIndex {
+					break
+				}
+				membership := memberships[membershipIndex]
+				if membership.LastViewedAt < postTimestamp && membership.LastViewedAt > postTimestamps[counter-1] {
+					params := map[string]interface{}{
+						"ChannelId":    channel.Id,
+						"UserId":       membership.UserId,
+						"MsgCountRoot": channel.TotalMsgCountRoot - int64(counter),
+					}
+					if _, err = transaction.Exec("UPDATE ChannelMembers SET MsgCountRoot=:MsgCountRoot WHERE ChannelId=:ChannelId AND UserId=:UserId", params); err != nil {
```

Comment:
I think it is ok to do it here, but we can squeeze a bit more performance by grouping all the memberships updated with the same count, and updating all with one query.

Reply:
Anyway... lets keep it as is now, and if we need that extra performance we can try to squeeze then, because I think the code is going to be a bit more hard to understand.